networking_aci/plugins/ml2/drivers/mech_aci/driver.py

```diff
@@ -279,24 +303,54 @@ def cleanup_segment_if_needed(self, context, port, network, binding_levels, segm
             return
 
         # Get segment from ml2_port_binding_levels based on segment_id and host
-        # if no ports on this segment for host we can remove the aci allocation
+        # if no ports on this segment for host we can remove the aci allocation.
+        # In baremetal mode we need to call cleanup when the hostgroup is no longer on the physnet
         released = self.allocations_manager.release_segment(network, hostgroup, 1, segment)
-        if not released:
+        if not released and not (hostgroup['direct_mode'] and hostgroup['hostgroup_mode'] == aci_const.MODE_BAREMETAL):
             return
 
         # Call to ACI to delete port if the segment is released i.e.
         # port is the last for the network one on the host
         # Check if physical domain should be cleared
-        clearable_physdoms = self._get_clearable_phys_doms(context, network['id'],
-                                                           segment, hostgroup)
-        self.rpc_notifier.delete_port(port, hostgroup, clearable_physdoms)
+        ACI_CONFIG.annotate_baremetal_info(hostgroup, network['id'], override_project_id=port['project_id'])
+        clearable_physdoms = []
+        if released:
+            # physdoms will only be removed on segment removal
+            clearable_physdoms = self._get_clearable_phys_doms(context, network['id'],
```

Comment:
An EPG will raise a fault if not associated with an AEP, right? We do not care about them? is there no order of operations that need to be maintained when attaching a VPC to an EPG, such as the EPG must have the AEP already?

Reply:
As discussed in slack, EPG <-> physdom is the relevant association. AEP is only the connection between VPC-policy and physdom.